### DISCUSSION TITLE
Adventure map, AI behavior compared to the original game, scenario DOMINION.MX2 (0.9.7 4306)

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#4325] [General]
In the original game, with this scenario, playing the Knight with the color blue, from the beginning of the game the human player is constantly harassed by the green AI player, by the yellow AI player and often by the orange AI player.
This scenario presents a real challenge in the original game and even more so when the scenario is played in impossible mode.

In fheroes2, it's really not the same.
The yellow AI player plays with his boats in his corner without ever coming to attack the blue human player.
The yellow AI player presents no challenge and is easily conquered as soon as the human player can recruit a large enough army from the Knight's castle and attack it from the sea.
The green AI player is also not on hand to harass the blue human player.
It turns out that AI players are conquered one after the other without any real resistance.

In the original game, the AI on the adventure map is much more aggressive, constantly harassing the human player from all sides.
If I remember correctly, the AI in the original game is not subject to the fog of war and can therefore see the entire map, it also has knowledge of whether a wandering troop would join it, the value of treasure chests, resource piles, and other things.
This allows the AI to have much better visibility to plan its moves and attacks on the entire map, not just the small part of the map it has "discovered".
Try this scenario in the original game in impossible mode and you will see the difference.

I don't know how this is currently handled in fheroes2, and I think the AI should have some advantages if it can allow it to play better strategically on the whole map.
What do you think about it?

**Comment by Branikolog** [comment]
Hello, @LeHerosInconnu.
If I'm not wrong, we currently haven't worked much on AI using boats or DD. This game mechanics requires advanced algorithm,  which prevents AI from wasting his time using boats or DD (I think AI doesn't know how to use DD at all now).
AI is so weak for you at this map exactly because of water nearby. Boarding and unboarding a boat requires a ton of movement points, so player should have strict plan while using boats. Try to play a map, where AI has minimum free access to a water. It can even be a challenge sometimes, to stand all his harassments. 
We have a lot of work with AI still. But I'm glad, your feedback is similar to mine. A problem being described is a step to fix it. ;)

**Comment by Gerwin2k** [comment]
I noticed a similar AI balance issue in the other direction, where it seems the AI is impossible to defeat. The stock Scenario "King of Death" with the default player setup. So in that case you start with the castle on the top left of the map, which is only a starting castle requiring the human player many turns to expand and upgrade it. 
However the computer possesses the castle in the bottom left of the map. This castle already has Bone Dragons ready for use. In the original game the AI tends to keep these bone dragons for defense. But in FHeroes2 the computer player transfers them to one or two of its heroes, and after like 10 turns goes on the offensive with them.

**Comment by oleg-derevenetz** [comment]
Hi @Gerwin2k I believe, all these issues described by @LeHerosInconnu should be not observed anymore as well. The AI's ability to use boats and teleports has improved significantly since 0.9.7, so AI should not waste time anymore even on large maps with water.

**Reply by Gerwin2k** [reply]
Good to hear. 

The issue I described is still there though, as I tested it again two days ago: Those bone dragons make any player effort futile.
But I understand that scenarios that rely on original passive AI behavior are a frustrating case.

**Reply by oleg-derevenetz** [reply]
@Gerwin2k That's precisely because AI is no longer wasting time. Maybe it would be possible to implement some sort of handicap for such cases (small map with a strong initial disparity in the strength of the troops) in future. I'm not a big expert in AI code, but I suppose that @idshibanov could have come up with something.

**Reply by Gerwin2k** [reply]
I hope so. 
I don't know how many such maps are with the game. 
(Small correction though, of my report for "King of Death": The computer starts with a Necromancer castle, with buildings for all dwellings in upgraded form. It does not have Bone Dragons yet, but can buy 1 at the start. The computer has plenty of gold to buy more once they are available.)

**Reply by idshibanov** [reply]
@Gerwin2k I just checked the scenario and can explain the difference. It has a "capture town" victory condition which forces original game AI to focus on defense and hold best army there doesn't matter if there's a real threat to it. Map design relies on this however it can cripple old AI's effectiveness on others (noticeable on OG campaign maps) even if it holds multiple castles.

Currently fheroes's AI does not look at map victory conditions at all. We already have an issue open for this #4418 . Once fixed this scenario should be easier, although I'm sure computer player won't behave exactly like in original.

**Reply by a1exsh** [reply]
I'm noticing a similar problem with the POL_2 scenario.  I've tried more than 10 times to win in it in fheroes2, but couldn't as the yellow player is too powerful at the beginning.  Now I see that he has Hydras pre-built in his remote castle on the island.  I'm constantly stumbling upon his heroes with packs of Hydras in the first weeks, which doesn't give me any chance.  In OG I could win from the first try...

**Reply by a1exsh** [reply]
Thought this time I could actually win, but nope: ![image](https://user-images.githubusercontent.com/489601/154853763-4de3e08b-6d35-49fc-8a5d-e3c49a397a41.png)

**Reply by a1exsh** [reply]
> Thought this time I could actually win, but nope

Wait, I've tricked AI by taking his undefended castles and buying his own Black Dragons while he was chasing my less useful heroes around %)  But it was still extremely hard compared to the OG experience.

**Reply by zenseii** [reply]
I've had the same experience playing the very same map, and I too had to exploit the AI to win. 

In my case I was playing an older release when the AI had issues deciding what to do on sea and therefore easy to abuse for that weakness. I'm afraid to try this map now after that behavior has been fixed....

One benefit/balance for the player in fheroes2 compared to the original is that the AI does not see the entire map, and this can be used to sneak up to one of the AI's better castles to launch a suprise attack. I'm not sure if there is some type of behavior specified in the map editor that the AI hasn't had implemented yet, like say a defensive mode rather than offensive explorative mode, but if so then that might make this map more feasible.

Trying to make the AI dumber just because the original maps were balanced for the original AI's shortcomings is not an option.

One option could be external patches for the maps, like xdelta patches, that change and rebalance these maps which have become unbalanced due to the AI's improved efficiency/competitiveness. I imagine these patches would have to be external to this project however.

**Comment by Gerwin2k** [comment]
I tried the "King of Death" scenario again with the latest SVN 0.9.19+. 

It seemed to play a little easier compared to past February, as the computer-opponent was initially not focused on attacking me. Though after about 10 turns that changed. 
I fooled the computer-opponent hero with bone dragons, by having my hero move in and out of my main castle. That kept the computer-opponent changing its direction repeatedly, essentially doing nothing useful with that unit. Until I had enough troops to actually defeat him, with heavy losses of my own. 
After that I had some hope of winning this scenario, but by that time the enemy had another hero with plenty of bone dragons coming, and I gave up. 

Tried some other tactics, basically with the same intention of keeping the computer-opponent wasting its time chasing a hero of mine. But the map is just too small for such. Especially once there are two strong enemy heroes. 

No surprise really, since the core of the matter has not changed in fheroes2, but I was curious about the small computer-opponent code changes of the past months.

**Reply by ihhub** [reply]
Hi @Gerwin2k ,  which game difficulty do you set? If you can't beat the map try the lower difficulty first.

**Reply by Gerwin2k** [reply]
Oh yeah, I had it set to "Easy" already. But forgot to mention that.

**Reply by ihhub** [reply]
And can you beat this map in the original game with no issues?

**Reply by zenseii** [reply]
I managed to beat King of Death on easy on the latest build after a couple of tries and I had to resort to some saving and reloading. I played barbarian and bought 2 heroes in the beginning and killed the first red hero with a main barbarian that had up to ogres. Killed off green pretty quickly and then it was mostly a game of cat and mouse. I realized I could lure the AI near the starting castle without them attacking, and they didn't seem interested in attacking green's castle either since they hadn't explored there so that gave me some time to accumulate some forces and a place to hide.

I mainly picked up gold from chests. Don't waste time on resources or items that you don't really need. In the end I managed to maneuver around the AI when the main hero was near my castle and then storming across the swamp towards the the Sherman castle and capturing it while the AI was busy doing other stuff.

If you manage to get to the genies then you could always use the halving tactic on them. I tested this as well when loading back to the point where I decided to make a run for it, and with the genies I was able to kill a total of 19 bone dragons and eventually decimate all the AI's forces.

I also did a playthrough, of Dominion on normal difficulty and did many mistakes. This was my first time trying this scenario. I never restarted nor loaded a save but still beat it without too much struggle. I think it would've been harder if the AI had been more eager to explore by sea.

**Reply by Gerwin2k** [reply]
@ihhub  Don't remember too well, I wrote this before.
"_This castle already has Bone Dragons ready for use. In the original game the AI tends to keep these bone dragons for defense._"
I will try the KoD scenario again with the original game in the weekend. 

@zenseii 
Zensei, thanks for the tips and the clarification that the KoD on Easy can indeed be won. I am just novice/average at this game, I will see if I can have more success with your tips in mind. 

> I also did a playthrough, of Dominion on normal difficulty and did many mistakes. This was my first time trying this scenario. I never restarted nor loaded a save but still beat it without too much struggle. I think it would've been harder if the AI had been more eager to explore by sea.

@zenseii note that this discussion mentions scenarios with different balance issues:  @LeHerosInconnu started by mentioning **Dominion** is **too easy** in FHeroes2, then in February I added that scenarios like **King of Death** were **impossibly hard** in FHeroes2. Since then AI perks were reduced so I decided to give it a try again past week.

**Reply by ihhub** [reply]
I'm honestly not surprised if playing a map in the original game is much easier. It's a well-known fact that AI there is pretty dumb.

**Reply by zenseii** [reply]
@Gerwin2k, yes I had never played any of these two maps in neither fheroes2 nor the original game (OG). From @LeHerosInconnu's description above the Dominion map has become harder and later I tested just letting the days pass without doing anything and yellow very quickly came and attacked with a boat.

The core difference between fheroes2 and the OG for large water maps such as Dominion is that the fheroes2 AI does not get the whole map revealed from the beginning, so this difference is more obvious in such a large map. However, I'm very confident the fheroes2 AI can become even more aggressive than the original when it becomes more efficient at exploring.

One thing I noticed was that the AI in fheroes2 does not prioritize staying on a road when exploring, which is not optimal since most map makers tend to create roads to lead you to the more important things like castles or the other teams. This is very evident in Dominion.

**Reply by Gerwin2k** [reply]
> I'm honestly not surprised if playing a map in the original game is much easier. It's a well-known fact that AI there is pretty dumb.

Of course. But they (new world computing) compensated for that in their scenario design, in the case of King of Death by giving the computer-opponent a fully upgraded starting castle. I figure, when the computer actually **uses** that advantage properly, it cannot lose.    
I have not played "Dominion".

**Reply by Gerwin2k** [reply]
> I will try the KoD scenario again with the original game in the weekend.

I just finished King of Death, in the original game, at normal difficulty. Only the red heroes which were pre-placed further from home are active. Two creatures blocking the red starting area were still there at at turn 50. So 4 red heroes were stuck in that area. 1 of the 25 red bone dragons ventured out of the castle, the rest stayed there waiting, until the human player decides he is ready to face them.

**Reply by zenseii** [reply]
It would be interesting to watch the AI play without fog of war on both OG and fheroes2.

**Comment by LeHerosInconnu** [comment]
I tested the DOMINION.MX2 scenario, blue player as Knight, with version 1.0.0.
The AI is still passive.
I won the scenario easily without challenge.
The AI never worried me and suffered my progress without real opposition.

**Comment by LeHerosInconnu** [comment]
This has improved with versions 1.0.8 7767 and 1.0.8 7846.
The AI is a little more aggressive on the adventure map.
The yellow AI player made a few attacks by sea on the blue human player's castles.
The green and red AI players have also taken some harassing actions against the blue human player.

Please note that this is the "DOMINION.MX2" scenario and not the "Dominion.mp2" scenario.

<details>
<summary> "DOMINION.MX2" PoL map: (click to expand)</summary>
<br>

![DOMINION (PoL) minimap](https://github.com/ihhub/fheroes2/assets/43583994/e2985ff4-6963-4c3e-89fe-c8fb876dd687)

</details>
<details>
<summary> "Dominion.mp2" SW map: (click to expand)</summary>
<br>

![Dominion (SW) minimap](https://github.com/ihhub/fheroes2/assets/43583994/70b26708-6694-475c-9a11-7003b8e4ca10)

</details>

**Comment by LeHerosInconnu** [comment]
These remarks are too old to be considered.
I close the discussion.
